### DISCUSSION TITLE
How to validate required_if on an enum attribute

### DISCUSSION THREAD

**Discussion by telkins** [spatie/laravel-data#244] [Q&A] [answered]
Hello,

I have a data class with a field that is type-hinted with a native PHP enum...something like this:
```php
#[MapInputName(SnakeCaseMapper::class)]
class MyDataClass extends Data
{
    public function __construct(
        public MyTypeEnum $type,
        #[RequiredIf('type', MyTypeEnum::SPECIAL_TYPE)]
        public ?string $myStringAttribute,
    ) {
    }
}
```

It doesn't seem to work.

I validate and create like so:
```php
$data = [
    'type' => MyTypeEnum::SPECIAL_TYPE,
    // 'my_string_attribute' => 'required, but missing',
];

$myDataObject = MyDataClass::from($data));
```

I have tried a all of the various ways I could think of:

- Using `MyTypeEnum::SPECIAL_TYPE->value` instead of `MyTypeEnum::SPECIAL_TYPE` for the rule.
- Moving the rule down to the static `rules()` method.
- Using Laravel's `Rule::requiredIf()` instead...which seems to only take a bool or a closure returning a bool but not getting the actual value.

I can't think of anything else to try.  I'm open to suggestions and would really welcome some good advice.

Thanks in advance.  🤓

**Comment by rubenvanassche** [comment, marked as answer]
Hi @telkins,

I don't know how you would get this working, since the RequiredIf rule does not support enums. I've added support for it and when I test my code it seems to be working. Please notice that `from` will never validate input use `validateAndCreate` instead.

Hopefully I can release it later today!

**Reply by telkins** [reply]
> Hi @telkins,
> 
> I don't know how you would get this working, since the RequiredIf rule does not support enums.

Well, I guess if it doesn't support enums, then that explains a few things.  I didn't realize that.  It was obvious that I couldn't get it to work, but I didn't know for sure if it was because of that or because I was doing something wrong.

> I've added support for it and when I test my code it seems to be working.

That's fantastic...!  I'll have to check it out.

In any case, is there any support for the after validation hook (https://laravel.com/docs/9.x/validation#after-validation-hook) or anything like that...? 🤔

> Please notice that `from` will never validate input use `validateAndCreate` instead.

Regarding `from()` vs `validateAndCreate()`...I realized that and I'm not sure why I wrote it that way in my example.  😅  In any case, thx for the note. 🤓

> Hopefully I can release it later today!

👍

**Reply by telkins** [reply]
@rubenvanassche Can you show me your PR/test(s) so that I can see what you've done.  I am completely up to date and it's still failing for me.  (And I'm using `validateAndCreate()`, just to be clear. 🤓 )

**Reply by rubenvanassche** [reply]
This works:

```php
it('bla', function(){
    #[MapInputName(SnakeCaseMapper::class)]
    class MyDataClass extends Data
    {
        public function __construct(
            public DummyBackedEnum $type,
            #[RequiredIf('type', DummyBackedEnum::FOO)]
            public ?string $myStringAttribute,
        ) {
        }
    }

    dd(MyDataClass::validate([
        'type' => DummyBackedEnum::FOO->value,
    ]));
});

```
